File: WADV.Core/Vector2.vb
''' <summary>
''' 表示一个二维向量
''' </summary>
Public Structure Vector2
    ''' <summary>
    ''' 向量的X轴分量
    ''' </summary>
    Public X As Double
    ''' <summary>
    ''' 向量的Y轴分量
    ''' </summary>
    Public Y As Double

    ''' <summary>
    ''' 获得一个新的二维向量
    ''' </summary>
    ''' <param name="x">向量的X轴分量</param>
    ''' <param name="y">向量的Y轴分量</param>
    Public Sub New(x As Double, y As Double)
        Me.X = x
        Me.Y = y
    End Sub

    Public Shared Operator +(a As Vector2, b As Vector2) As Vector2
        Return New Vector2(a.X + b.X, a.Y + b.Y)
    End Operator

    Public Shared Operator +(a As Vector2, b As Double) As Vector2
        Return New Vector2(a.X + b, a.Y + b)
    End Operator

    Public Shared Operator +(a As Vector2, b As Integer) As Vector2
        Return New Vector2(a.X + b, a.Y + b)
    End Operator

    Public Shared Operator +(a As Vector2, b As Point2) As Vector2
        Return New Vector2(a.X + b.X, a.Y + b.Y)
    End Operator

    Public Shared Operator -(a As Vector2, b As Vector2) As Vector2
        Return New Vector2(a.X - b.X, a.Y - b.Y)
    End Operator

    Public Shared Operator -(a As Vector2, b As Double) As Vector2
        Return New Vector2(a.X - b, a.Y - b)
    End Operator

    Public Shared Operator -(a As Vector2, b As Integer) As Vector2
        Return New Vector2(a.X - b, a.Y - b)
    End Operator

    Public Shared Operator -(a As Vector2, b As Point2) As Vector2
        Return New Vector2(a.X - b.X, a.Y - b.Y)
    End Operator

    Public Shared Operator *(a As Vector2, b As Vector2) As Double
        Return a.X * b.X + a.Y * b.Y
    End Operator

    Public Shared Operator *(a As Vector2, b As Double) As Vector2
        Return New Vector2(a.X * b, a.Y * b)
    End Operator

    Public Shared Operator *(a As Vector2, b As Integer) As Vector2
        Return New Vector2(a.X * b, a.Y * b)
    End Operator

    Public Shared Operator *(a As Vector2, b As Point2) As Vector2
        Return New Vector2(a.X * b.X, a.Y * b.X)
    End Operator

    Public Shared Operator /(a As Vector2, b As Double) As Vector2
        Return New Vector2(a.X / b, a.Y / b)
    End Operator

    Public Shared Operator /(a As Vector2, b As Integer) As Vector2
        Return New Vector2(a.X / b, a.Y / b)
    End Operator

    Public Shared Operator /(a As Vector2, b As Point2) As Vector2
        Return New Vector2(a.X / b.X, a.Y / b.X)
    End Operator

    Public Shared Operator =(a As Vector2, b As Vector2) As Boolean
        Return a.X = b.X AndAlso a.Y = b.Y
    End Operator

    Public Shared Operator =(a As Vector2, b As Double) As Boolean
        Return [Mod](a) = b
    End Operator

    Public Shared Operator <>(a As Vector2, b As Vector2) As Boolean
        Return Not a = b
    End Operator

    Public Shared Operator <>(a As Vector2, b As Double) As Boolean
        Return Not a = b
    End Operator

    Public Shared Operator >(a As Vector2, b As Vector2) As Boolean
        Return (a.X * a.X + a.Y * a.Y) > (b.X * b.X + b.Y * b.Y)
    End Operator

    Public Shared Operator <(a As Vector2, b As Vector2) As Boolean
        Return (a.X * a.X + a.Y * a.Y) < (b.X * b.X + b.Y * b.Y)
    End Operator

    Public Shared Operator >=(a As Vector2, b As Vector2) As Boolean
        Return (a.X * a.X + a.Y * a.Y) >= (b.X * b.X + b.Y * b.Y)
    End Operator

    Public Shared Operator <=(a As Vector2, b As Vector2) As Boolean
        Return (a.X * a.X + a.Y * a.Y) <= (b.X * b.X + b.Y * b.Y)
    End Operator

    Public Shared Widening Operator CType(v As Integer) As Vector2
        Return New Vector2(v, v)
    End Operator

    Public Shared Widening Operator CType(v As Double) As Vector2
        Return New Vector2(v, v)
    End Operator

    ''' <summary>
    ''' 求取二维向量的模
    ''' </summary>
    ''' <param name="target">要使用的二维向量</param>
    ''' <returns></returns>
    Public Shared Function [Mod](target As Vector2) As Double
        Return Math.Sqrt(Mod2(target))
    End Function

    ''' <summary>
    ''' 求取二维向量的模的平方
    ''' </summary>
    ''' <param name="target">要使用的二维向量</param>
    ''' <returns></returns>
    Public Shared Function Mod2(target As Vector2) As Double
        Return target.X * target.X + target.Y * target.Y
    End Function

    ''' <summary>
    ''' 零向量
    ''' </summary>
    ''' <returns></returns>
    Public Shared ReadOnly Property Zero As Vector2 = New Vector2 With {.X = 0.0, .Y = 0.0}

    ''' <summary>
    ''' 单位向量
    ''' </summary>
    ''' <returns></returns>
    Public Shared ReadOnly Property Unit As Vector2 = New Vector2 With {.X = 1.0, .Y = 1.0}
End Structure

File: WADV.Core.NETCore/Utility/Vector2.cs
using System;

namespace WADV.Core.NETCore.Utility
{
    /// <summary>
    /// 2D vector
    /// </summary>
    public struct Vector2
    {
        /// <summary>
        /// X component
        /// </summary>
        public double X;
        /// <summary>
        /// Y component
        /// </summary>
        public double Y;

        /// <summary>
        /// Get a new 2D vector
        /// </summary>
        /// <param name="x">X component</param>
        /// <param name="y">Y component</param>
        public Vector2(double x, double y)
        {
            X = x;
            Y = y;
        }

        public static Vector2 operator +(Vector2 a, Vector2 b)
        {
            return new Vector2(a.X + b.X, a.Y + b.Y);
        }
        public static Vector2 operator +(Vector2 a, double b)
        {
            return new Vector2(a.X + b, a.Y + b);
        }
        public static Vector2 operator +(Vector2 a, int b)
        {
            return new Vector2(a.X + b, a.Y + b);
        }
        public static Vector2 operator +(Vector2 a, Point2 b)
        {
            return new Vector2(a.X + b.X, a.Y + b.Y);
        }
        public static Vector2 operator -(Vector2 a, Vector2 b)
        {
            return new Vector2(a.X - b.X, a.Y - b.Y);
        }
        public static Vector2 operator -(Vector2 a, double b)
        {
            return new Vector2(a.X - b, a.Y - b);
        }
        public static Vector2 operator -(Vector2 a, int b)
        {
            return new Vector2(a.X - b, a.Y - b);
        }
        public static Vector2 operator -(Vector2 a, Point2 b)
        {
            return new Vector2(a.X - b.X, a.Y - b.Y);
        }
        public static double operator *(Vector2 a, Vector2 b)
        {
            return a.X * b.X + a.Y * b.Y;
        }
        public static Vector2 operator *(Vector2 a, double b)
        {
            return new Vector2(a.X * b, a.Y * b);
        }
        public static Vector2 operator *(Vector2 a, int b)
        {
            return new Vector2(a.X * b, a.Y * b);
        }
        public static Vector2 operator *(Vector2 a, Point2 b)
        {
            return new Vector2(a.X * b.X, a.Y * b.X);
        }
        public static Vector2 operator /(Vector2 a, double b)
        {
            return new Vector2(a.X / b, a.Y / b);
        }
        public static Vector2 operator /(Vector2 a, int b)
        {
            return new Vector2(a.X / b, a.Y / b);
        }
        public static Vector2 operator /(Vector2 a, Point2 b)
        {
            return new Vector2(a.X / b.X, a.Y / b.X);
        }
        public static bool operator ==(Vector2 a, Vector2 b)
        {
            return a.X.Equals(b.X) && a.Y.Equals(b.Y);
        }
        public static bool operator ==(Vector2 a, double b)
        {
            return Mod(a).Equals(b);
        }
        public static bool operator !=(Vector2 a, Vector2 b)
        {
            return !(a == b);
        }
        public static bool operator !=(Vector2 a, double b)
        {
            return !(a == b);
        }
        public static bool operator >(Vector2 a, Vector2 b)
        {
            return (a.X * a.X + a.Y * a.Y) > (b.X * b.X + b.Y * b.Y);
        }
        public static bool operator <(Vector2 a, Vector2 b)
        {
            return (a.X * a.X + a.Y * a.Y) < (b.X * b.X + b.Y * b.Y);
        }
        public static bool operator >=(Vector2 a, Vector2 b)
        {
            return (a.X * a.X + a.Y * a.Y) >= (b.X * b.X + b.Y * b.Y);
        }
        public static bool operator <=(Vector2 a, Vector2 b)
        {
            return (a.X * a.X + a.Y * a.Y) <= (b.X * b.X + b.Y * b.Y);
        }
        public static implicit operator Vector2(int v)
        {
            return new Vector2(v, v);
        }
        public static implicit operator Vector2(double v)
        {
            return new Vector2(v, v);
        }

        /// <summary>
        /// Get the mod of this vector
        /// </summary>
        /// <returns></returns>
        public double Mod()
        {
            return Math.Sqrt(Mod2());
        }
        /// <summary>
        /// Get the square mod of this vector
        /// </summary>
        /// <returns></returns>
        public double Mod2()
        {
            return X * X + Y * Y;
        }

        /// <summary>
        /// Zero vector
        /// </summary>
        /// <returns></returns>
        public static Vector2 Zero { get; } = new Vector2(0, 0);
        /// <summary>
        /// Unit vector
        /// </summary>
        /// <returns></returns>
        public static Vector2 Unit { get; } = new Vector2(1, 1);

        public bool Equals(Vector2 other)
        {
            return X.Equals(other.X) && Y.Equals(other.Y);
        }
        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            return obj is Vector2 && Equals((Vector2)obj);
        }
        public override int GetHashCode()
        {
            unchecked
            {
                return (X.GetHashCode() * 397) ^ Y.GetHashCode();
            }
        }
    }
}